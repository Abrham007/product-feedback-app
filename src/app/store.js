import { configureStore } from "@reduxjs/toolkit";

import productRequestsReducer from "../features/productRequests/productRequestsSlice";

export default configureStore({
  reducer: {
    productRequests: productRequestsReducer,
  },
});
