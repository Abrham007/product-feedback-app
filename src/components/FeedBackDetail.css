.FeedBackDetail {
  width: 730px;
  display: flex;
  gap: 24px;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin: 80px 0px 120px;
}

.FeedBackDetail__header {
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.FeedBackDetail__link {
  display: flex;
  gap: 4px;
  align-items: center;
  border: none;
  outline: none;
  flex-shrink: 0;
  border-radius: 10px;
  padding: 12px 24px;
  color: #f2f4fe;
  font-size: 0.875rem;
  font-weight: 700;
  text-decoration: none;
}

.FeedBackDetail__link--1 {
  gap: 15px;
  color: #647196;
  background-color: transparent;
  padding: 0px;
}

.FeedBackDetail__link--1:hover {
  cursor: pointer;
  text-decoration: underline;
}

.FeedBackDetail__link--2 {
  background-color: #4661e6;
}

.FeedBackDetail__link--2:hover {
  cursor: pointer;
  background-color: #7c91f9;
}

@media screen and (max-width: 768px) {
  .FeedBackDetail {
    width: 327px;
    margin: 24px 0px 88px;
  }

  .FeedBackDetail__btn {
    padding: 10.5px 16px;
    font-size: 0.8125rem;
  }
}
