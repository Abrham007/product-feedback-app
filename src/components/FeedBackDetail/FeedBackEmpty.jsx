import "./FeedBackEmpty.css";
import emptyIcon from "../../assets/suggestions/illustration-empty.svg";

import AddBtn from "../UI/Buttons/AddBtn";
import { useSelector } from "react-redux";

function FeedBackEmpty() {
  const postStatus = useSelector((state) => state.productRequests.status);

  // We set up a loading state for the loading stage
  let isLoading = false;
  if (postStatus === "loading") {
    isLoading = true;
  }

  return (
    <div className="FeedBackEmpty ">
      <img
        className="FeedBackEmpty__icon"
        src={emptyIcon}
        alt=""
        width={130}
        height={136}
      ></img>
      <div className="FeedBackEmpty__text">
        <h2 className="FeedBackEmpty__header">
          {isLoading ? "Fethcing feedback..." : "There is no feedback yet."}
        </h2>
        <p className="FeedBackEmpty__paragraph">
          {isLoading
            ? "This may take a few seconds"
            : "Got a suggestion? Found a bug that needs to be squashed? We love hearing about new ideas to improve our app."}
        </p>
      </div>
      <AddBtn />
    </div>
  );
}

export default FeedBackEmpty;
