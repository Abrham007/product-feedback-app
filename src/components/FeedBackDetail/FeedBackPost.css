.FeedBackPost {
  width: 100%;
  height: 151px;
  display: flex;
  gap: 40px;
  padding: 28px 32px;
  border-radius: 10px;
  background: #fff;
}

.FeedBackPost:hover {
  cursor: pointer;
}

.FeedBackPost:hover .FeedBackPost__header {
  color: #4661e6;
}

.FeedBackPost__text {
  height: 100%;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  gap: 4px;
  margin-right: auto;
}

.FeedBackPost__header {
  color: #3a4374;
  font-size: 1.125rem;
  font-weight: 700;
  letter-spacing: -0.01563rem;
}

.FeedBackPost__paragraph {
  color: #647196;
  font-size: 1rem;
  font-weight: 400;
  margin-bottom: auto;
}

.FeedBackPost__comments {
  align-self: center;
}

@media screen and (max-width: 768px) {
  .FeedBackPost {
    width: 327px;
    height: 200px;
    display: grid;
    gap: 16px;
    grid-template-columns: 1fr 1fr;
    grid-template-rows: 4fr 1fr;
    grid-template-areas:
      "text text"
      "score comments";
    padding: 24px;
  }

  .FeedBackPost__score {
    grid-area: score;
  }

  .FeedBackPost__text {
    grid-area: text;
  }

  .FeedBackPost__comments {
    grid-area: comments;
    justify-self: flex-end;
  }

  .FeedBackPost__header {
    font-size: 0.8125rem;
    letter-spacing: -0.01131rem;
  }

  .FeedBackPost__paragraph {
    font-size: 0.8125rem;
  }
}
