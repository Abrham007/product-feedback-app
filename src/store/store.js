import { configureStore } from "@reduxjs/toolkit";
import currentUserReducer from "../features/currentUser/currentUserSlice";
import productRequestsReducer from "../features/productRequests/productRequestsSlice";

export default configureStore({
  reducer: {
    currentUser: currentUserReducer,
    productRequests: productRequestsReducer,
  },
});
