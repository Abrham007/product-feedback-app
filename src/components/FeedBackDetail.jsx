import "./FeedBackDetail.css";
import FeedBackPost from "./FeedBackPost";
import leftArrowIcon from "../assets/shared/icon-arrow-left.svg";
import CommentList from "./CommentList";
import AddComment from "./AddComment";
import { Link, useParams } from "react-router-dom";

function FeedBackDetail(props) {
  let { id } = useParams();

  const FEEDBACKDETAIL = props.appData.productRequests.find(
    (req) => req.id == id
  );

  const FEEDBACKDETAILCOMMENTS = FEEDBACKDETAIL.comments;

  return (
    <div className="FeedBackDetail">
      <div className="FeedBackDetail__header">
        <Link to="/" className="FeedBackDetail__link FeedBackDetail__link--1">
          <svg width="7" height="10" xmlns="http://www.w3.org/2000/svg">
            <path
              d="M6 9L2 5l4-4"
              stroke="#647196"
              strokeWidth="2"
              fill="none"
              fillRule="evenodd"
            />
          </svg>
          <span>Go Back</span>
        </Link>
        <button className="FeedBackDetail__link FeedBackDetail__link--2">
          <span>Edit Feedback</span>
        </button>
      </div>
      <FeedBackPost {...FEEDBACKDETAIL} />
      <CommentList
        list={FEEDBACKDETAILCOMMENTS}
        handleAppData={props.handleAppData}
        id={id}
      />
      <AddComment id={id} handleAppData={props.handleAppData} />
    </div>
  );
}

export default FeedBackDetail;
