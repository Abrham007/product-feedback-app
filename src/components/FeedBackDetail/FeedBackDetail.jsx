import "./FeedBackDetail.css";
import FeedBackPost from "./FeedBackPost";
import CommentList from "../Comments/CommentList";
import AddComment from "../Comments/AddComment";
import { Link, useParams } from "react-router-dom";
import BackBtn from "../UI/Buttons/BackBtn";
import { useSelector } from "react-redux";
function FeedBackDetail() {
  // This is the id we get from either the suggesitons page or
  // the roadmap page
  let { id } = useParams();

  // We filter the post to show the show the post with the specific _id
  const FEEDBACKDETAIL = useSelector((state) =>
    state.productRequests.posts.find((req) => req._id == id)
  );

  // we isolate the comment of that post if it exist or it is undefiend
  const FEEDBACKDETAILCOMMENTS = FEEDBACKDETAIL?.comments;

  return (
    <div className="FeedBackDetail">
      <div className="FeedBackDetail__header">
        <BackBtn />
        <Link to={`/edit/${id}`} className="FeedBackDetail__link">
          Edit Feedback
        </Link>
      </div>
      <FeedBackPost {...FEEDBACKDETAIL} />
      {FEEDBACKDETAILCOMMENTS && (
        <CommentList list={FEEDBACKDETAILCOMMENTS} id={id} />
      )}

      <AddComment id={id} />
    </div>
  );
}

export default FeedBackDetail;
