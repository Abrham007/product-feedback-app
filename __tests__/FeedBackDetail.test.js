import FeedBackDetail from "../src/components/FeedbackDetail";
import { Route, MemoryRouter, Routes } from "react-router-dom";
import { preloadedData } from "../utils/test-utils";
import { setupStore } from "../src/store/store";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";

describe("FeedBackDetail ", () => {
  const component = (
    <Provider store={setupStore(preloadedData)}>
      <MemoryRouter initialEntries={[`/feedbackdetail/${2}`]}>
        <Routes>
          <Route
            path="/feedbackdetail/:id"
            element={<FeedBackDetail></FeedBackDetail>}
          ></Route>
        </Routes>
      </MemoryRouter>
    </Provider>
  );
  test("renders the correct JSX ", () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
