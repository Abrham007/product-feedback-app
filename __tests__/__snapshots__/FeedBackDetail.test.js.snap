// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedBackDetail  renders the correct JSX  1`] = `
<div
  className="FeedBackDetail"
>
  <div
    className="FeedBackDetail__header"
  >
    <button
      className="BackBtn"
      onClick={[Function]}
      style={
        {
          "color": "#647196",
        }
      }
    >
      <svg
        height="10"
        width="7"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 9L2 5l4-4"
          fill="none"
          fillRule="evenodd"
          stroke="#4661E6"
          strokeWidth="2"
        />
      </svg>
      <span>
        Go Back
      </span>
    </button>
    <a
      className="FeedBackDetail__link"
      href="/edit/2"
      onClick={[Function]}
    >
      Edit Feedback
    </a>
  </div>
  <a
    className="FeedBackPost"
    href="/feedbackdetail/2"
    onClick={[Function]}
    style={
      {
        "textDecoration": "none",
      }
    }
  >
    <div
      className="FeedBackPost__score"
    >
      <button
        className="ScoreButton"
        disabled={false}
        onClick={[Function]}
        style={
          {
            "cursor": "pointer",
          }
        }
      >
        <svg
          height="7"
          style={
            {
              "stroke": "#4661E6",
            }
          }
          width="10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 6l4-4 4 4"
            fill="none"
            fillRule="evenodd"
            strokeWidth="2"
          />
        </svg>
        <span>
          99
        </span>
      </button>
    </div>
    <div
      className="FeedBackPost__text"
    >
      <h3
        className="FeedBackPost__header"
      >
        Add a dark theme option
      </h3>
      <p
        className="FeedBackPost__paragraph"
      >
        It would help people with light sensitivities and who prefer dark mode.
      </p>
      <button
        className="Tag"
        disabled={true}
        onClick={[Function]}
        style={
          {
            "backgroundColor": "#f2f4ff",
            "cursor": "text",
          }
        }
      >
        feature
      </button>
    </div>
    <div
      className="FeedBackPost__comments"
    >
      <div
        className="PostComments"
      >
        <img
          alt=""
          height={16}
          width={18}
        />
        <p
          style={{}}
        >
          2
        </p>
      </div>
    </div>
  </a>
  <div
    className="CommentList"
  >
    <h4
      className="CommentList__header"
    >
      2
       Comments
    </h4>
    <div
      className="CommentPost"
    >
      <img
        alt=""
        className="CommentPost__avatar"
        height={40}
        src="file:///C:/Users/abrha/Desktop/FrontEnd%20Mentor/Guru/product-feedback/product-feedback-app/src/assets/user-images/image-elijah.jpg"
        width={40}
      />
      <div
        className="CommentPost__info"
      >
        <h5
          className="CommentPost__name"
        >
          Elijah Moss
        </h5>
        <p
          className="CommentPost__username"
        >
          @
          hexagon.bestagon
        </p>
      </div>
      <button
        className="CommentPost__btn-get"
        onClick={[Function]}
      >
        Replay
      </button>
      <p
        className="CommentPost__description"
      >
        <span
          className="CommentPost__replayto"
        />
        Also, please allow styles to be applied based on system preferences. I would love to be able to browse Frontend Mentor in the evening after my device’s dark mode turns on without the bright background it currently has.
      </p>
    </div>
    <hr
      className="CommentList__line"
    />
    <div
      className="CommentPostWithReplay"
    >
      <hr
        className="CommentPostWithReplay__line"
        style={
          {
            "height": "266px",
          }
        }
      />
      <div
        className="CommentPostWithReplay__comment"
      >
        <div
          className="CommentPost"
        >
          <img
            alt=""
            className="CommentPost__avatar"
            height={40}
            src="file:///C:/Users/abrha/Desktop/FrontEnd%20Mentor/Guru/product-feedback/product-feedback-app/src/assets/user-images/image-james.jpg"
            width={40}
          />
          <div
            className="CommentPost__info"
          >
            <h5
              className="CommentPost__name"
            >
              James Skinner
            </h5>
            <p
              className="CommentPost__username"
            >
              @
              hummingbird1
            </p>
          </div>
          <button
            className="CommentPost__btn-get"
            onClick={[Function]}
          >
            Replay
          </button>
          <p
            className="CommentPost__description"
          >
            <span
              className="CommentPost__replayto"
            />
            Second this! I do a lot of late night coding and reading. Adding a dark theme can be great for preventing eye strain and the headaches that result. It’s also quite a trend with modern apps and  apparently saves battery life.
          </p>
        </div>
      </div>
      <div
        className="CommentPostWithReplay__replies"
      >
        <div
          className="CommentPost"
        >
          <img
            alt=""
            className="CommentPost__avatar"
            height={40}
            src="file:///C:/Users/abrha/Desktop/FrontEnd%20Mentor/Guru/product-feedback/product-feedback-app/src/assets/user-images/image-anne.jpg"
            width={40}
          />
          <div
            className="CommentPost__info"
          >
            <h5
              className="CommentPost__name"
            >
              Anne Valentine
            </h5>
            <p
              className="CommentPost__username"
            >
              @
              annev1990
            </p>
          </div>
          <button
            className="CommentPost__btn-get"
            onClick={[Function]}
          >
            Replay
          </button>
          <p
            className="CommentPost__description"
          >
            <span
              className="CommentPost__replayto"
            >
              @hummingbird1 
            </span>
            While waiting for dark mode, there are browser extensions that will also do the job. Search for 'dark theme' followed by your browser. There might be a need to turn off the extension for sites with naturally black backgrounds though.
          </p>
        </div>
        <div
          className="CommentPost"
        >
          <img
            alt=""
            className="CommentPost__avatar"
            height={40}
            src="file:///C:/Users/abrha/Desktop/FrontEnd%20Mentor/Guru/product-feedback/product-feedback-app/src/assets/user-images/image-ryan.jpg"
            width={40}
          />
          <div
            className="CommentPost__info"
          >
            <h5
              className="CommentPost__name"
            >
              Ryan Welles
            </h5>
            <p
              className="CommentPost__username"
            >
              @
              voyager.344
            </p>
          </div>
          <button
            className="CommentPost__btn-get"
            onClick={[Function]}
          >
            Replay
          </button>
          <p
            className="CommentPost__description"
          >
            <span
              className="CommentPost__replayto"
            >
              @annev1990 
            </span>
            Good point! Using any kind of style extension is great and can be highly customizable, like the ability to change contrast and brightness. I'd prefer not to use one of such extensions, however, for security and privacy reasons.
          </p>
        </div>
      </div>
    </div>
    <hr
      className="CommentList__line"
    />
  </div>
  <form
    className="AddComment"
    onSubmit={[Function]}
  >
    <label
      className="AddComment__header"
      htmlFor="comment"
    >
      Add Comment
    </label>
    <textarea
      className="AddComment__input"
      id="comment"
      maxLength={250}
      name="comment"
      onChange={[Function]}
      placeholder="Type your comment here"
      value=""
    />
    <div
      className="AddComment__box"
    >
      <p
        className="AddComment__char"
      >
        250
         Characters left
      </p>
      <button
        className="AddComment__btn"
        type="submit"
      >
        Post Comment
      </button>
    </div>
  </form>
</div>
`;
